File: hook/gate_overrides.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
)

// gate_overrides.go mirrors packages/core/src/gate-overrides.ts --
// ~/.twing/gate-overrides.json, a per-project map deciding whether the §17
// design gate is disabled for that project on this machine. Exists because
// hook wiring is machine-global now (wire-hooks.ts) instead of per-repo, so
// `twing design disable-gate` can no longer work by unwiring a repo-local
// hook entry -- unwiring a global entry would disable the gate everywhere,
// not just one repo. Read-only on this side -- only the TS CLI
// (`design enable-gate`/`disable-gate`) ever writes this file.
func gateOverridesPath() (string, bool) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", false
	}
	return filepath.Join(home, ".twing", "gate-overrides.json"), true
}

func isGateDisabled(projectID string) bool {
	path, ok := gateOverridesPath()
	if !ok {
		return false
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return false
	}
	var overrides map[string]string
	if err := json.Unmarshal(data, &overrides); err != nil {
		return false
	}
	return overrides[projectID] == "disabled"
}

File: packages/core/src/gate-overrides.ts
/**
 * Per-repo override for the §17 design gate, decoupled from hook-entry
 * wiring now that wiring is machine-global (`wire-hooks.ts`) instead of
 * per-repo (§ install-once-per-machine onboarding work). `enable-gate`/
 * `disable-gate` used to work by wiring/unwiring `PreToolUse`/`SessionEnd`
 * entries in a *repo-local* `.claude/settings.json` -- that stopped making
 * sense once wiring is global (unwiring a global entry would disable the
 * gate everywhere, not just one repo). This is the local, per-project,
 * machine-local replacement: a flat map, mirroring `config.ts`'s
 * per-server auth map pattern. Read by the Go hook (`hook/gate_overrides.go`,
 * a read-only mirror -- only this side ever writes it) on every real gate
 * check.
 */

import * as fs from "node:fs";
import * as os from "node:os";
import * as path from "node:path";

interface GateOverrides {
  [projectId: string]: "disabled";
}

function overridesPath(): string {
  return path.join(os.homedir(), ".twing", "gate-overrides.json");
}

function readOverrides(): GateOverrides {
  try {
    return JSON.parse(fs.readFileSync(overridesPath(), "utf8")) as GateOverrides;
  } catch {
    return {};
  }
}

function writeOverrides(overrides: GateOverrides): void {
  const target = overridesPath();
  fs.mkdirSync(path.dirname(target), { recursive: true });
  fs.writeFileSync(target, JSON.stringify(overrides, null, 2) + "\n");
}

export function isGateDisabled(projectId: string): boolean {
  return readOverrides()[projectId] === "disabled";
}

export function setGateDisabled(projectId: string, disabled: boolean): void {
  const overrides = readOverrides();
  if (disabled) {
    overrides[projectId] = "disabled";
  } else {
    delete overrides[projectId];
  }
  writeOverrides(overrides);
}
